packages/Kokkos/src/details/DTK_DetailsTreeTraversal.hpp

```diff
@@ -3,101 +3,46 @@
 
 namespace DataTransferKit
 {
-
-struct Node;
-
 namespace Details
 {
 
-// priority queue helper for nearest neighbor search
-using Value = std::pair<Node const *, double>;
-
-struct CompareDistance
-{
-    bool operator()( Value const &lhs, Value const &rhs )
-    {
-        // larger distance means lower priority
-        return lhs.second > rhs.second;
-    }
-};
-
-using PriorityQueue =
-    std::priority_queue<Value, std::vector<Value>, CompareDistance>;
-
-// helper for k nearest neighbors search
-// container to store candidates throughout the search
-struct SortedList
-{
-    using Value = std::pair<int, double>;
-    using Container = std::list<Value>;
-    SortedList( int k )
-        : _maxsize( k ){};
-    template <typename... Args>
-    void emplace( Args &&... args )
-    {
-        if ( _sorted_list.size() < _maxsize )
-            _sorted_list.emplace_back( std::forward<Args>( args )... );
-        else
-        {
-            Value val( std::forward<Args>( args )... );
-            if ( !_compare( val, _sorted_list.back() ) )
-                throw std::runtime_error(
-                    "Error in DTK::Impl::SortedList::emplace()" );
-            _sorted_list.back() = val;
-        }
-        _sorted_list.sort( _compare );
-    }
-    bool empty() const { return _sorted_list.empty(); }
-    bool full() const { return _sorted_list.size() == _maxsize; }
-    Container::size_type size() const { return _sorted_list.size(); }
-    Value const &back() const { return _sorted_list.back(); }
-    Container _sorted_list;
-    // closer objects stored at the front of the list
-    std::function<bool( Value const &a, Value const &b )> _compare =
-        []( Value const &a, Value const &b ) { return a.second < b.second; };
-    Container::size_type _maxsize;
-};
-
 // There are two (related) families of search: one using a spatial predicate and
 // one using nearest neighbours query (see boost::geometry::queries
 // documentation).
 template <typename NO, typename Predicate>
 KOKKOS_FUNCTION void
 spatial_query( BVH<NO> const bvh, Predicate const &predicate, int *indices,
-               unsigned int &n_indices, unsigned int n_max_indices )
+               unsigned int &n_indices, unsigned int max_n_indices )
 {
-    // Allocate traversal stack from thread-local memory, and push nullptr to
-    // indicate that there are no postponed nodes.
-    Node const *stack[64];
-    Node const **stack_ptr = stack;
-    *stack_ptr++ = nullptr;
+    Stack<Node const *> stack;
 
     Node const *node = BVHQuery<NO>::getRoot( bvh );
-    unsigned int pos = 0;
-    do
+    stack.push( node );
+    n_indices = 0;
+
+    while ( !stack.empty() )
     {
+        node = stack.top();
+        stack.pop();
+
         if ( BVHQuery<NO>::isLeaf( node ) )
         {
-            indices[pos] = BVHQuery<NO>::getIndex( bvh, node );
-            ++pos;
 #if HAVE_DTK_DBC
-            if ( pos > n_max_indices )
+            if ( n_indices > max_n_indices )
```

Comment:
I have thought more about this and we need an `assert` after the `printf`. On the GPU, the code will crash right away but not on the CPU, so we need the `assert`.